### ISSUE TITLE
Use color class system instead of `display-graphic-p` for Daemon

### ISSUE THREAD

**Issue by mekeor** [greduan/emacs-theme-gruvbox#54]
I'm opening a new issue because the OP of #37 was originally describing a different issue, namely the error output while the theme was loaded correctly which is not the case for me.

# Description:

- Launch Emacs Daemon within terminal emulator within graphical desktop environment: `emacs --daemon`.
- Launch an Emacs Client within the terminal emulator: `emacsclient -c`.
- The theme is not loaded (correctly). For instance, the background-color is still white. Also, I get the same `Unable to load color "color-NNN"` error messages as in #37 .

[Here is a screenshot.](http://imgur.com/a/4cR1r)

# Explanation (from an [Emacs-StackExchange-Answer](http://emacs.stackexchange.com/a/26138/13390))

> No nearest match translation is done. (The visual results are bad[.]) themes have a color class system which they are supposed to use, instead of the display-graphic-p solution used in Gruvbox. 

> The only problems with the display-graphic-p solution is that it only calculates at first load. Which can cause issues when running an Emacs daemon and starting both GUI and Terminal emacsclient.

# Proposed Solution:

Let's use the **color class system** instead of `display-graphic-p`. I could port it myself but I need more information about it. Searching for "emacs color class system" doesn't help. Can you point me to some direction?

**Comment by mekeor** [comment]
Is the color class system what is described here? https://www.gnu.org/software/emacs/manual/html_node/elisp/Defining-Faces.html

**Comment by jasonm23** [comment]
By the way the background color being wrong AND the Unable to load color message are not actually related.

The Unable to load color message is simply a spurious error, the xterm terminal (or emulator) will still get the color name sent, and map it as per xterm.

See this chart for a list of xterm color names.

https://en.m.wikipedia.org/wiki/File:Xterm_256color_chart.svg

**Comment by jasonm23** [comment]
Slight error on my part, that chart lists the xterm color numbers .

They are named as `color-nnn`

**Comment by jasonm23** [comment]
By the way the color class entries are horribly verbose...

```
(defface highlight
  '((((class color) (min-colors 88) (background light))
     :background "darkseagreen2")
    (((class color) (min-colors 88) (background dark))
     :background "darkolivegreen")
    (((class color) (min-colors 16) (background light))
     :background "darkseagreen2")
    (((class color) (min-colors 16) (background dark))
     :background "darkolivegreen")
    (((class color) (min-colors 8))
     :background "green" :foreground "black")
    (t :inverse-video t))
  "Basic face for highlighting."
  :group 'basic-faces)
```

**Comment by mekeor** [comment]
Today, I tried to understand and apply the color class syntax. In fact, it's a bit redundant, so I'd like to ask you if you have an idea how to get rid of the redundancy.

Here's the theme I'm currently testing on: https://gist.github.com/mekeor/a4ed7c7439fdf761dddcef5d5c20ed76 (I'm just experimenting – I'm not planning to publish another gruvbox theme for Emacs.)

**Comment by jasonm23** [comment]
Look at Autothemer 

https://github.com/sebastiansturm/autothemer

**Comment by jasonm23** [comment]
FYI I'll be releasing 0.2.0 of Darktooth using Autothemer later today.

It won't be on master branch, so MELPA users of Darktooth won't suffer from any migration issues until I've dogfooded it sufficiently.

More details soon.  Autothemer must be manually installed from file.

**Comment by jasonm23** [comment]
https://github.com/emacsfodder/emacs-theme-darktooth/tree/use-autothemer

**Comment by jasonm23** [comment]
also added autothemer to my dev work on creamsody, it's working very well.

**Comment by jasonm23** [comment]
I'll convert Gruvbox tomorrow.

**Comment by mekeor** [comment]
Are you on it? Btw, are you on IRC so we can communicate faster? I've got a working version of gruvbox theme using autothemer now.

**Comment by mekeor** [comment]
It's not finished yet but it works: https://github.com/mekeor/emacs-theme-gruvbox/blob/autothemer/gruvbox-theme.el

**Comment by jasonm23** [comment]
Please check out the autothemer-amends branch.

BTW due to this significant slice of work, I'm adding you as a named author, please set the email.

Thanks for the great contribution.

**Comment by mekeor** [comment]
Cool. Sure. Thanks, too. I will add more modes the following days. :)

**Comment by jasonm23** [comment]
Please hold off on any more additions, if you have anything to merge right now, please rebase on the autothemer-amends.

(or just do a diff and manual patch if you see a lot of conflicts.)
I can do this later on today probably if you don't have time.

I don't want to add any more modes really, not until after we've got the autothemer version released.

This way we can have smaller, atomic commits. They are more useful patches than a huge commit like this.

When doing what amounts to a reformat, it's best not to introduce a bunch of other changes until we are working on  specific patches for them.

**Comment by jasonm23** [comment]
I've re-done the autothemer port from scratch because I had to integrate a couple of new PRs.

(Also after upgrading two other themes to use autothemer, I've got a quick workflow / macros etc. to make it a five minute task.)

I've released it to be built by MELPA and MELPA stable, so I hope we can put a few open issues to bed now.

@mekeor - check out `autothemer-generate-templates` if you haven't already.

The next  big task is to integrate the alternative palettes.
